### ISSUE TITLE
Integrating with Openresty

### ISSUE THREAD

**Issue by TafkaMax** [coreruleset/modsecurity-crs-docker#117]
Hi

I would like to know, if you are open to creating an openresty version of the modsec-crs-docker?

https://kevalnagda.github.io/configure-nginx-and-keycloak-to-enable-sso-for-proxied-applications

I am looking into creating a NGINX proxy, that is the relying party for an keycloak server and all the guides use Openresty, as the base, but I would like to have the functionalities of MODSEC and the CRS.

**Comment by theseion** [comment]
Interesting idea. Unfortunately, we have enough on our hands to maintain the current set of images. As we use those images ourselves we also regularly test at least some of the features. Adding something like openresty would bloat the image further and we wouldn't be testing that it works, as we don't use it. That's my personal opinion at least. @fzipi?

Feel free, of course, to use our images as base for an openresty image that you publish yourself.

**Comment by TafkaMax** [comment]
Hmm I am trying to test this.

Currently I tried to replace NGINX base image with Openresty. 

I might change this around to use openresty buil with modsec-crs base image.

CUrrently I am seeing long modsecurity build times using 

```bash
# I added a new version to build
docker buildx bake -f docker-bake.hcl openresty-alpine-fat
```

----

This command, that is given in your README gives me an ERROR, not adding it seems to work.
```bash
docker buildx bake -f ./docker-bake.hcl --set target.platforms=linux/amd64 openresty-alpine-fat
[+] Building 0.0s (0/0)                                                                                                                                                                                                                                                                   
ERROR: could not find any target matching 'target'
```


BTW I tried to use

**Comment by TafkaMax** [comment]
I can build the image, but launching it seems to be problematic:

Something i have to figure out.

```
#docker logs openresty-testing -f
Error relocating /usr/local/openresty/bin/openresty: SSL_magic_pending_session_ptr: symbol not found
```

**Comment by fzipi** [comment]
Hi @TafkaMax!

I have some questions about the OpenResty integration you want to do.
1. why only `linux/amd64`? we support way many archs in other versions, this looks like a step back
2. If we don't change the configuration from nginx, why not using the same configuration? We can use a new variable like `$prefix` to change between `nginx` and `openresty`, if necessary.
3. I don't know if I want to be the maintainer right now. Can you be the one maintaining this, as we don't normally use OpenResty in our daily lives, it will simplify acceptance.
4. Why not using https://github.com/gogatekeeper/gatekeeper or https://github.com/oauth2-proxy/oauth2-proxy instead?

**Comment by fzipi** [comment]
@TafkaMax Friendly ping!

**Comment by TafkaMax** [comment]
Hi.

1. I was also able to build it for linux/arm64. I haven't looked deeply into this, but I think the Openresty base image is built for 64bit only, which is the reason why I was not able to build it for arm32/v7 and i386. I need to look into how the Openresty image is built and maybe it can be built for 32 bit aswell. If not, I hope supporting both 64bit systems of linux/amd64 and arm64 is OK from your point of view?
2. I think I don't wholly understand what you mean with this. Could you please elaborate?
3. I need to think about this, but overall I think I can maintain the Openresty part.
4. I haven't looked at those webservers. Thank you for sharing. My main reason why to use openresty is so that I can use the WAF  functionality of modsec with the core rule set. Does gatekeeper and oauth2-proxy support Modsecurity?

Sorry for the delay in response.

**Comment by TafkaMax** [comment]
I tested building for 32bit systems. It seems the Openresty image I use as the base image is not built for 32bit systems.

```
ERROR: failed to solve: openresty/openresty:1.21.4.1-alpine-fat: no match for platform in manifest sha256:a0e1c2d5ea07c586cfc3b1ec83383406a52711f7c782f2a471e31aeaaebcb2f0: not found
```

I tried to also create a debian base Docker container, but adding the packages I wished  to add(Luarocks), it became a little too complicated, so I decided to not include the Dockerfile for the Debian based container for now.

**Comment by theseion** [comment]
Hi @TafkaMax. I'm sorry that we haven't responded. We're still under water with v4.0.

Are there any missing pieces from your side?
@fzipi I think you need to elaborate one question.

**Comment by TafkaMax** [comment]
After looking at my proposed changes I can see I have some cleaning up to do. 

1. docker-compose.yml I have no intention of overwriting the previous changes. 
2. .gitlab-ci.yml could be left out. But if somebody forks this repo, it is a nice base where to start from. It doesn't automatically build all the 18 possibilities, you can choose which architecture and system bit type you wish to build.
3. Because nginx is the reverse proxy openresty uses. I can just copy the files that are the same from `nginx/templates` and `nginx/docker-entrypoint.d` inside the Dockerfile, currently I have just copied them over to openresty folder. But some of the files I can't just copy and paste. For example `nginx.conf` is different.

**Comment by theseion** [comment]
Let us know, when you're ready.

**Comment by TafkaMax** [comment]
OK I have updated the README.md and added information regarding Openresty.

I have also reworked some of the includes in the Dockerfile. So everything that is shared between the nginx version and openresty version that is the same, is copied from the `nginx` folder. That includes most of the `templates` folder, except for the `nginx.conf.template`.

I have reverted the docker-compose.yml back to the way it is in the upstream repo.

If you go over this and do not wish to integrate these changes. It is fine by me As you originally said: 

> Feel free, of course, to use our images as base for an openresty image that you publish yourself.

I myself am building and using the openresty versions of modsecurity-crs in my companies infrastructure for half a year now. My main reason was to share the knowledge to others and also if these changes get commited, then I can download the images from dockerhub instead from the private repositories where I am currently building them.
